orttraining/orttraining/core/optimizer/triton_fusion.cc

```diff
@@ -0,0 +1,432 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#ifdef ENABLE_TRITONOP
+
+#include <nlohmann/json.hpp>
+#include <string_view>
+
+#include "orttraining/core/optimizer/triton_fusion.h"
+
+#include "core/framework/compute_capability.h"
+#include "core/graph/model.h"
+#include "core/graph/graph_utils.h"
+#include "core/optimizer/initializer.h"
+#include "core/optimizer/utils.h"
+#include "core/providers/partitioning_utils.h"
+
+using namespace ONNX_NAMESPACE;
+using json = nlohmann::json;
+
+namespace onnxruntime {
+
+namespace {
+
+using SizeTypeVec = InlinedVector<size_t>;
+using NodeVec = InlinedVector<Node*>;
+using NodeArgVec = InlinedVector<NodeArg*>;
+using ConstNodeArgVec = InlinedVector<const NodeArg*>;
+using NodeArgSet = InlinedHashSet<NodeArg*>;
+using IsSupportedFunc = std::function<bool(const Graph&, const Node&)>;
+
+int64_t Hash(const std::string& str) {
+  uint32_t hash = 0;
+  for (char const& c : str) {
+    hash = hash * 101 + c;
+  }
+
+  return static_cast<int64_t>(hash);
+}
+
+bool CheckAxis(const Node& node, int64_t expected_axis) {
+  const auto& attributes = node.GetAttributes();
+  if (attributes.find("axis") == attributes.end() || !attributes.at("axis").has_i()) return false;
+  int64_t axis = attributes.at("axis").i();
+  if (axis == expected_axis) return true;
+  if (axis < 0 || expected_axis < 0) {
+    const auto& input_shape = node.InputDefs()[0]->Shape();
+    if (input_shape == nullptr) return false;
+    int64_t rank = static_cast<int64_t>(input_shape->dim_size());
+    if (axis < 0) axis += rank;
+    int64_t non_neg_expected_axis = expected_axis < 0 ? expected_axis + rank : expected_axis;
+    return axis == non_neg_expected_axis;
+  }
+  return false;
+}
+
+bool CheckAxes(const Graph& graph, const Node& node, bool single_axis, const std::vector<int>& expected_axes) {
+  std::vector<int64_t> axes_values;
+  const auto& attributes = node.GetAttributes();
+  if (attributes.find("axes") != attributes.end()) {
+    axes_values = RetrieveValues<int64_t>(attributes.at("axes"));
+  } else if (node.InputDefs().size() == 2) {
```

Comment:
Can we alwasy assume the input counts are equal to 2? Or should we rename `CheckAxes` to CheckReduceNodeAxes

Reply:
We can't. Even for OpSet >= 13, missing 2nd input means reduce on all axes by default.
Axes attribute/input is not just for reduce nodes, we can also apply this rule check for other kinds of nodes with this attribute/input if we want to support them in the future.